[WIP] RAIL notification icon support

hualet opened the pull request:
Now the implementation can work as:

1. show a tray icon for remote apps.
2. support basic button clicks.

but it also have known issues:

1. the icon is poorly drawn, the background is black.
2. the local tray icon windows sometimes are not embeded into the system tray.
3. right click on the tray icon, the context menu shows at somewhere else.

trying to fix: #5108

akallabeth commented on client/X11/xf_rail.c:
```diff
@@ -698,6 +708,23 @@ static BOOL convert_icon_color_to_argb(ICON_INFO* iconInfo, BYTE* argbPixels)
 }
 
+static void convert_argb_to_bgra(BYTE* argbPixels, BYTE* bgraPixels, int len)
+{
```
Why not use `freerdp_image_copy`?

akallabeth commented on client/X11/xf_window.c:
```diff
@@ -1133,3 +1132,84 @@ xfAppWindow* xf_AppWindowFromX11Window(xfContext* xfc, Window wnd)
+	int count;
+	ULONG_PTR* pKeys = NULL;
+	xfAppNotifyIcon* notifyIcon;
+	count = HashTable_GetKeys(xfc->railNotifyIcons, &pKeys);
```
`xfc->railNotifyIcons == NULL` if not used with RAIL

akallabeth commented on client/X11/xf_rail.c:
```diff
@@ -880,6 +907,35 @@ static BOOL xf_rail_window_cached_icon(rdpContext* context,
+
+	if (orderInfo->fieldFlags & WINDOW_ORDER_STATE_NEW)
+	{
+		notifyIcon = (xfAppNotifyIcon*) calloc(1, sizeof(xfAppNotifyIcon));
```
unchecked allocation

akallabeth commented on client/X11/xf_rail.c:
```diff
@@ -1214,6 +1311,17 @@ static void rail_window_free(void* value)
+		return;
+
+	XUnmapWindow(icon->xfc->display, icon->handle);
+	XDestroyWindow(icon->xfc->display, icon->handle);
```
memory leak, `xfAppNotifyIcon` was allocated with `calloc`

akallabeth commented on client/X11/xf_rail.c:
```diff
@@ -880,6 +907,35 @@ static BOOL xf_rail_window_cached_icon(rdpContext* context,
+		if (!notifyIcon)
+			return FALSE;
+
+		HashTable_Add(xfc->railNotifyIcons, (void*)(UINT_PTR) orderInfo->notifyIconId,
```
'0' is a valid notification ID but `HashTable_Add` will not accept that.

hualet replied:
Does that situation that '0' as a notification ID exist in practical ? How do I suppose to fix this ? :sweat_smile:

akallabeth replied:
yes it does. @hardening pointed me in a simple direction, just set a comparison function that compares the pointers by type (UINT32) and use something like `&notifyIcon->notifyIconId` as key

hardening commented:
@akallabeth could be nice to report the crash conditions to fedora (as an X client is not supposed to crash the whole server)

akallabeth commented:
@hardening still trying to get a proper crash dump ;)

akallabeth commented:
The tray icons should be set by `libappindicator` or similar and not as a simple X window if possible.

hualet commented:
Thank you for your reviewing, I'll update the PR soon.

> The tray icons should be set by `libappindicator` or similar and not as a simple X window if possible.

The reason that I didn't use libappindicator is that I don't know if it's proper to introduce new dependencies :P, if that's acceptable, I'll try that @akallabeth

hualet commented:
Oh, libappindicator seems to be only working with GTK,  so I don't think it's a very good choice.

akallabeth commented:
@hualet problem with x11 windows is that they look like the app has 2 entries in most environments then but still no notification icon. 
New deps are no problem, just check it still works without them activated ;)

hualet commented:
> New deps are no problem, just check it still works without them activated ;)

I'm reading gtk's source code to see how can I fix the tray icon problem.
because introducing libappindicator/gtk means we need another main loop, really overwhelming I think @akallabeth

akallabeth commented on client/X11/xf_rail.c:
```diff
@@ -955,40 +948,29 @@ static BOOL xf_rail_notify_icon_common(rdpContext* context,
 
 		notifyIcon->image = image;
-		free(argbPixels);
+		// bgraPixels will be freed by XDestroyImage.
```
No C++ comments please

hualet commented:
@akallabeth I've fixed all of the mistakes you mentioned above, except:

1. hashtable don't accept uint32, which I've created a new PR https://github.com/FreeRDP/FreeRDP/pull/5120 trying to follow up the direction you told me.
2. the tray icon background is black, I still don't have a very good way to fix that.

can you review my code again, thanks?

akallabeth commented:
@hualet 
* The hash table issue can be solved just like #5117 
* I'll have a look at it.

akallabeth commented:
@hualet I've fixed a few remaining issues and rebased to current master. See https://github.com/akallabeth/FreeRDP/tree/rail_notify_icons (note: I squashed your changes to a single commit)

hualet commented:
> @hualet I've fixed a few remaining issues and rebased to current master. See https://github.com/akallabeth/FreeRDP/tree/rail_notify_icons (note: I squashed your changes to a single commit)

OK, so I need to close this PR now?

akallabeth commented:
@hualet if you want to. I'm still struggling with getting the notification icon to work though.
I can see the window icon in the window bar (gnome that is with the extension for it) but:
* No icon for the notification
* Everything I expect from the notification icon is not working, aka context menu, ...

hualet commented:
> @hualet if you want to. I'm still struggling with getting the notification icon to work though.
> I can see the window icon in the window bar (gnome that is with the extension for it) but:
> 
> * No icon for the notification
> * Everything I expect from the notification icon is not working, aka context menu, ...

Hmmm.... I tested that with Mate, DDE and stalonetray, all seems good,  can you test with stalonetray ?

hualet commented:
Another thing I must mention is that the context menu is not working perfectly here too, the context menu shows at random position :(

akallabeth commented:
@hualet I've tried with cinnamon and mate and only in mate I could see an `icon` (with a width of 1 pixel)

hualet commented:
> @hualet I've tried with cinnamon and mate and only in mate I could see an `icon` (with a width of 1 pixel)

Which application are you testing with? I can test it here too.

akallabeth commented:
@hualet Yes, with KDE plasma I can see an icon too. Application not relevant, all behave the same in this regard.
Still, I get many problems with nearly all WM so this will need some work before getting merged.
